lib/api/api-fetch/response.js

```diff
@@ -1,141 +1,348 @@
+  static error () {
+    // The static error() method steps are to return the result of creating a
+    // Response object, given a new network error, "immutable", and this’s
+    // relevant Realm.
+    // TODO: relevant Realm?
+    const responseObject = new Response()
+    responseObject[kState] = makeNetworkError()
+    responseObject[kHeaders][kHeadersList] = responseObject[kState].headersList
+    responseObject[kHeaders][kGuard] = 'immutable'
+    return responseObject
+  }
+
+  // Creates a redirect Response that redirects to url with status status.
+  static redirect (url, status = 302) {
+    // 1. Let parsedURL be the result of parsing url with current settings
+    // object’s API base URL.
+    // 2. If parsedURL is failure, then throw a TypeError.
+    // TODO: base-URL?
+    let parsedURL
+    try {
+      parsedURL = new URL(url)
+    } catch (err) {
+      const error = new TypeError()
+      error.cause = err
+      throw error
     }
+
+    // 3. If status is not a redirect status, then throw a RangeError.
+    if (!redirectStatus.includes(status)) {
+      throw new RangeError()
+    }
+
+    // 4. Let responseObject be the result of creating a Response object,
+    // given a new response, "immutable", and this’s relevant Realm.
+    // TODO: relevant Realm?
+    const responseObject = new Response()
+    responseObject[kHeaders][kGuard] = 'immutable'
+
+    // 5. Set responseObject’s response’s status to status.
+    responseObject[kState].status = status
+
+    // 6. Let value be parsedURL, serialized and isomorphic encoded.
+    // TODO: isomorphic encoded?
+    const value = parsedURL.toString()
+
+    // 7. Append `Location`/value to responseObject’s response’s header list.
+    responseObject[kState].headersList.push('location', value)
+
+    // 8. Return responseObject.
+    return responseObject
   }
 
+  // https://fetch.spec.whatwg.org/#dom-response
+  constructor (body = null, init = {}) {
+    // 1. If init["status"] is not in the range 200 to 599, inclusive, then
+    // throw a RangeError.
+    if ('status' in init) {
+      if (!Number.isFinite(init.status)) {
+        throw new TypeError()
+      }
+
+      if (init.status < 200 || init.status > 599) {
+        throw new RangeError()
+      }
+    }
+
+    if ('statusText' in init) {
+      if (typeof init.statusText !== 'string') {
+        throw new TypeError()
+      }
+
+      // 2. If init["statusText"] does not match the reason-phrase token
+      // production, then throw a TypeError.
+      // See, https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.2:
+      //   reason-phrase  = *( HTAB / SP / VCHAR / obs-text )
+      // TODO
+    }
+
+    // 3. Set this’s response to a new response.
+    this[kState] = makeResponse(init)
+
+    // 4. Set this’s headers to a new Headers object with this’s relevant
+    // Realm, whose header list is this’s response’s header list and guard
+    // is "response".
+    // TODO: relevant Realm?
```

Comment:
Realm might be browser environment only 🤷‍♂️

Reply:
I think realm is going to be part of es spec. So we might get it in node as well.